core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackApiViewService.java

```diff
@@ -80,34 +78,53 @@ public StackViewV4Response retrieveById(Long stackId) {
         }
     }
 
-    private Set<StackViewV4Response> filterDatalakes(boolean dataLakeOnly, Set<StackViewV4Response> stackViewResponses) {
+    public Environment decorate(Environment environment, Long workspaceId) {
+        ShowTerminatedClustersAfterConfig showTerminatedClustersAfterConfig = showTerminatedClusterConfigService.get();
+        if (showTerminatedClustersAfterConfig.isActive()) {
+            environment.setStacks(getAllByWorkspaceAndEnvironment(workspaceId, environment.getId(), showTerminatedClustersAfterConfig));
+        }
+        return environment;
+    }
+
+    private Set<StackApiView> filterDatalakes(boolean dataLakeOnly, Set<StackApiView> stackViewResponses) {
         if (dataLakeOnly) {
-            stackViewResponses = stackViewResponses
-                    .stream()
+            stackViewResponses = stackViewResponses.stream()
                     .filter(stackViewResponse ->
-                            Boolean.TRUE.equals(stackViewResponse.getCluster().getClusterDefinition().getTags().get("shared_services_ready")))
+                            Boolean.TRUE.equals(stackViewResponse.getCluster().getClusterDefinition().getTags().getMap().get("shared_services_ready")))
                     .collect(Collectors.toSet());
         }
         return new HashSet<>(stackViewResponses);
     }
 
-    private Set<StackViewV4Response> getAllByWorkspaceAndEnvironment(Long workspaceId, String environmentName) throws TransactionExecutionException {
+    private Set<StackApiView> getAllByWorkspaceAndEnvironment(Long workspaceId, String environmentName,
+            ShowTerminatedClustersAfterConfig showTerminatedClustersAfter) {
         EnvironmentView env = environmentViewService.getByNameForWorkspaceId(environmentName, workspaceId);
```

Comment:
How about a repository and service method like: `getEnvironmentIdByName`. Even more lightweight than the view.